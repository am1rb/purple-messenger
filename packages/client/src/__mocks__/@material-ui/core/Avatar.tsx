import React from "react";
import type { AvatarProps } from "@material-ui/core/Avatar";

function MuiAvatar({ src, alt }: AvatarProps) {
  return <img data-testid="mock-mui-avatar" src={src} alt={alt} />;
}

export default MuiAvatar;
