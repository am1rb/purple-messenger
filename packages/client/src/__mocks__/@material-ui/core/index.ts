export { default as Avatar } from "./Avatar";
export { default as Box } from "./Box";
export { default as Card } from "./Card";
export { default as CardContent } from "./CardContent";
export { default as Typography } from "./Typography";
export { default as Button } from "./Button";
