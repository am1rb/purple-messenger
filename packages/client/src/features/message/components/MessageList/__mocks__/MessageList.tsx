import React from "react";

const messageList = jest.fn(() => <div data-testid="mock-message-list" />);

export default messageList;
