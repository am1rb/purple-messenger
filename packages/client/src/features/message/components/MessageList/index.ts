export { default } from "./MessageList";
export * from "./MessageList";
