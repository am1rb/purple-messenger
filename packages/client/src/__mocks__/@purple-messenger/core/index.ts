import * as purpleMessengerCore from "@purple-messenger/core";

const mockedPurpleMessengerCore = jest.genMockFromModule<
  typeof purpleMessengerCore
>("@purple-messenger/core");

function mockedAction(action = () => ({ type: "" })) {
  return jest.fn(action);
}

module.exports = {
  ...mockedPurpleMessengerCore,
  loadConversationList: mockedAction(),
  unloadConversationList: mockedAction(),
  signOut: mockedAction(),
  clearMessageList: mockedAction(),
  seenMessageAck: mockedAction(),
  start: mockedAction(),
  shutdown: mockedAction(),
  submitMessage: mockedAction(),
  startTypingMessage: mockedAction(),
  stopTypingMessage: mockedAction(),
  signIn: mockedAction(),
  setAuthError: mockedAction(),
  verifyToken: mockedAction(),
  connected: mockedAction(purpleMessengerCore.connected),
  disconnected: mockedAction(purpleMessengerCore.disconnected),
};
