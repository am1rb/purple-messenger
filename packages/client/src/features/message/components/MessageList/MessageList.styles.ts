import { makeStyles } from "@material-ui/core/styles";

export default makeStyles({
  root: {
    overflowX: "hidden",
    overflowY: "scroll",
    height: "100%",
  },
});
