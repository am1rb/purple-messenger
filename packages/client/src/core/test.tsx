import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { ConnectedRouter } from "connected-react-router";
import configureStore, { history } from "core/redux/configureStore";
import { State as RootState } from "core/redux/reducers";
import theme from "core/style/theme";
import { ThemeProvider } from "@material-ui/styles";

export function renderWithStore(
  ui: React.ReactElement,
  initStore: Partial<RootState>
) {
  const store = configureStore(initStore);

  return render(ui, {
    wrapper: ({ children }) => (
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <ConnectedRouter history={history}>{children}</ConnectedRouter>
        </Provider>
      </ThemeProvider>
    ),
  });
}
