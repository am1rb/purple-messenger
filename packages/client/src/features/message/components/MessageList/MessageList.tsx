import React, {
  memo,
  useEffect,
  useRef,
  useMemo,
  useLayoutEffect,
} from "react";
import { useSelector, useDispatch } from "react-redux";
import { getMessageList } from "features/message/selectors";
import MessageRow from "../MessageRow";
import useConversationInfo from "features/conversation/components/useConversationInfo";
import { clearMessageList } from "@purple-messenger/core";
import { mustScrollToTheEnd } from "../../helpers";
import useStyles from "./MessageList.styles";

function MessageList() {
  const classes = useStyles();
  const dispatch = useDispatch();
  const messages = useSelector(getMessageList);
  const { username } = useConversationInfo();
  const messageListRef = useRef<HTMLDivElement>(null);
  const lastMessageRef = useRef<HTMLDivElement>(null);
  const needScroll = useRef(true);

  useEffect(() => {
    dispatch(clearMessageList());
  }, [username]);

  needScroll.current = useMemo(() => {
    const scrollbar = messageListRef.current;
    return scrollbar ? mustScrollToTheEnd(scrollbar, messages) : false;
  }, [messages]);

  useLayoutEffect(() => {
    if (needScroll.current) {
      lastMessageRef.current?.scrollIntoView();
    }
  }, [messages]);

  return (
    <div className={classes.root} ref={messageListRef}>
      {messages.map((message) => (
        <MessageRow key={message.id} {...message} />
      ))}
      <div ref={lastMessageRef} />
    </div>
  );
}

export default memo(MessageList);
