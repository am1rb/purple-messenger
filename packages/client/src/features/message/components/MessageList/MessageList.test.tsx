import React from "react";
import { OrderedMap } from "immutable";
import { renderWithStore, InitialStore } from "core/test";
import { sampleMessageList } from "@purple-messenger/core";
import { clearMessageList } from "@purple-messenger/core";
import MessageList from "./MessageList";

jest.mock("../MessageRow");

const initStore: InitialStore = {
  message: {
    lastMessageId: -1,
    list: OrderedMap(sampleMessageList.map((message) => [message.id, message])),
  },
};

describe("The <MessageList /> tests", () => {
  it("Should match the snapshot", () => {
    const { container } = renderWithStore(
      <MessageList username="john" />,
      initStore
    );
    expect(container).toMatchSnapshot();
  });

  it("Should call clearMessageList when the username prop is changed", () => {
    const { rerender, store } = renderWithStore(
      <MessageList username="john" />,
      initStore
    );

    expect(store.actions).toContainEqual(clearMessageList());

    store.clearActions();
    rerender(<MessageList username="sara" />);

    expect(store.actions).toContainEqual(clearMessageList());
  });
});
