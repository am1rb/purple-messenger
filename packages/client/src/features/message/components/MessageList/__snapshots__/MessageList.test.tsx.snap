// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <MessageList /> tests Should match the snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <div
      data-testid="mock-message-row"
    />
    <div
      data-testid="mock-message-row"
    />
    <div
      data-testid="mock-message-row"
    />
    <div />
  </div>
</div>
`;
